### PR TITLE
[libc] Add functions to send/recv messages

### PR DESCRIPTION
**Pull request by michaelrj-google**
This patch adds the necessary functions to send and receive messages
over a socket. Those functions are: recv, recvfrom, recvmsg, send,
sendto, sendmsg, and socketpair for testing.

### CODE REVIEW

**Comment by michaelrj-google** [comment]
This patch is incomplete currently. The functions are there, but I couldn't get the tests working. I'm uploading this here in case anyone wants to help with that. The tests only need to make sure each of these functions passes and fails, since we're only testing the syscall wrapper and not the actual kernel.

File: libc/include/llvm-libc-types/struct_iovec.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_TYPES_STRUCT_IOVEC_H__
+#define __LLVM_LIBC_TYPES_STRUCT_IOVEC_H__
+
+#include "llvm-libc-types/size_t.h"
```
Comment (lntue) [inline libc/include/llvm-libc-types/struct_iovec.h:12]: it should be relative to this file's path, so just:
```
#include "size_t.h"
```

File: libc/include/llvm-libc-types/struct_msghdr.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_TYPES_STRUCT_MSGHDR_H__
+#define __LLVM_LIBC_TYPES_STRUCT_MSGHDR_H__
+
+#include "llvm-libc-types/size_t.h"
```
Comment (lntue) [inline libc/include/llvm-libc-types/struct_msghdr.h:12]: The include path here should be relative to the file.

File: libc/include/llvm-libc-types/struct_msghdr.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
+//===-- Definition of struct msghdr ---------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_TYPES_STRUCT_MSGHDR_H__
```
Comment (lntue) [inline libc/include/llvm-libc-types/struct_msghdr.h:9]: no pre/post `__` in the header guards.

File: libc/include/llvm-libc-types/struct_iovec.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//===-- Definition of struct iovec ----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __LLVM_LIBC_TYPES_STRUCT_IOVEC_H__
```
Comment (lntue) [inline libc/include/llvm-libc-types/struct_iovec.h:9]: no pre/post `__` in the header guards.

File: libc/test/src/sys/socket/linux/send_recv_test.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+TEST(LlvmLibcSendRecvTest, RecvFails) {
+  char buffer[256];
+
+  ssize_t recv_result = LIBC_NAMESPACE::recv(-1, buffer, sizeof(buffer), 0);
+  ASSERT_EQ(recv_result, ssize_t(-1));
+  ASSERT_ERRNO_FAILURE();
+
+  LIBC_NAMESPACE::libc_errno = 0; // reset errno to avoid test ordering issues.
```
Comment (lntue) [inline libc/test/src/sys/socket/linux/send_recv_test.cpp:72]: Shall we also reset the `libc_errno` to 0 at the beginning of these tests?

Comment (michaelrj-google) [reply libc/test/src/sys/socket/linux/send_recv_test.cpp:72]: right now we assume that all tests start with `libc_errno = 0`. At some point we should enforce that in our test framework, but that's out of scope for this patch.

**Review by lntue — APPROVED** [review]

**Pull request merged**